applications/incompressible_navier_stokes/orr_sommerfeld/include/perturbation_energy.h

```diff
@@ -89,28 +93,35 @@ class PerturbationEnergyCalculator
       clear_files(true),
       initial_perturbation_energy_has_been_calculated(false),
       initial_perturbation_energy(1.0),
-      matrix_free_data(nullptr),
+      matrix_free(nullptr),
       dof_index(0),
       quad_index(0)
   {
   }
 
   void
-  setup(MatrixFree<dim, Number> const & matrix_free_data_in,
+  setup(MatrixFree<dim, Number> const & matrix_free_in,
         unsigned int const              dof_index_in,
         unsigned int const              quad_index_in,
         PerturbationEnergyData const &  data_in)
   {
-    matrix_free_data = &matrix_free_data_in;
-    dof_index        = dof_index_in;
-    quad_index       = quad_index_in;
-    energy_data      = data_in;
+    matrix_free = &matrix_free_in;
+    dof_index   = dof_index_in;
+    quad_index  = quad_index_in;
+    data        = data_in;
+
+    if(data.calculate)
+    {
+      // create directory if not already existing
+      if(Utilities::MPI::this_mpi_process(mpi_comm) == 0)
+        std::filesystem::create_directories(data.directory);
```

Comment:
Do you think we should check the return value of `create_directories` that tells us that the directory was created? We could otherwise throw an assert that we could not create (which might happen when the user gives a directory it does not have permission to write into)?

Reply:
I did not think about this. Maybe I assumed that one uses only paths relative to the ExaDG directory (where permissions should not be an issue?), but this is not necessarily the case. In principle, we could of course do something like that. In this case I would introduce a small utility function in ExaDG with `directory` and `mpi_comm` as parameters. We should first agree on how this looks like exactly, so that I have to go through all the changes only once more.

Reply:
I agree that a utility function that centralizes these two lines plus captures the result would be good. And as to the rights: You never know how a user sets up the things, so any guidance on possible errors might be helpful.

Reply:
I introduced a utility function here https://github.com/exadg/exadg/pull/62/commits/2e02391ebfc841e01c906be0610ab79fed19645e

Reply:
@kronbichler Now we could introduce the error handling also easily in a separate PR. Maybe this would be easier because I don't know what you think of exactly.

Reply:
I checked the new code with an invalid directory and I think the messages are good enough, so no need for further action:
```
Setup incompressible Navier-Stokes operator ...

... done!
terminate called after throwing an instance of 'std::filesystem::__cxx11::filesystem_error'
  what():  filesystem error: cannot create directories: Permission denied [/output/flow_past_cylinder/]
[mklap5:04334] *** Process received signal ***
[mklap5:04334] Signal: Aborted (6)
[mklap5:04334] Signal code:  (-6)
```

Reply:
This comes from the automatic exception that gets raised.